### ISSUE TITLE
Persistence not supported

### ISSUE THREAD

**Issue by PGijsbers** [facultyai/dash-bootstrap-components#283]
As far as I can tell, at this moment the Dash 1.3 feature [persistence](https://dash.plot.ly/persistence) is not supported. Are there plans for supporting this feature and if so, is there any timeline? I realize this would potentially raise the minimum Dash version requirement.

**Comment by tcbegley** [comment]
Hi @PGijsbers 

Thanks for raising this. I hadn't actually registered that this was something we could support. Looks like it's not too hard on our end so we should do it. Not too worried about bumping the version requirement.

Are there particular components / props that you would most like to have persisted?

**Comment by PGijsbers** [comment]
Cool! `Checklist` and `Input` are most important to me right now.

**Comment by tcbegley** [comment]
Waiting on [this issue](https://github.com/plotly/dash/issues/1034) to be resolved before making the required changes here.

**Comment by tcbegley** [comment]
Added preliminary support for this in 0.8.0. Currently persistence is supported for `Checklist`, `Input`, `RadioItems`, `Select`, `Tabs` and `Textarea`. When the above linked issue is resolved I will add support for more components. Will leave this issue open for now to keep tracking.

**Comment by Lemming0815** [comment]
could persistence also be supported for stuff like `Collapse`?

**Comment by tcbegley** [comment]
Yep, that's the plan. The reason it isn't yet is that it was not previously possible to persist boolean properties like `is_open` of `Collapse`. It looks like that might have been resolved by [this recently merged pull request](https://github.com/plotly/dash/pull/1142), so when a new release of Dash is made I hope to be able to support it.

**Comment by tcbegley** [comment]
Having looked into it in a bit more detail I'm not sure that PR will be enough. Might need some more changes to Dash still. Will probably be a little while before I get another chance to take a look.

**Comment by eddy-geek** [comment]
My PR to persist boolean properties has been released in Dash 1.10.0, in case that helps :-)
I believe you can customize what you persist as well, so, not sure what you ean, can you expand on this?

>  I'm not sure that PR will be enough. Might need some more changes to Dash still.

**Comment by tcbegley** [comment]
Thanks @eddy-geek,

I tried briefly to add support for the remaining components when I saw your changes got merged. I had a few issues (which could very possibly be me doing something dumb) but I didn’t get a chance to look into it further.

I plan to give it another go and I’ll be sure to let you know how I get on.

**Comment by MM-Lehmann** [comment]
Can you please check persistence against the new style `id=dict(...) `property? It doesn't seem to work with dash1.12. When I convert back to standard `id=str(...),` it does work.

**Comment by tcbegley** [comment]
@MM-Lehmann I can take a look for sure. Does persistence work with dash-core-components and `id=dict(...)`?

**Comment by MM-Lehmann** [comment]
@tcbegley good question. It doesn't actually. So this seems a dash issue rather. I'll post a new issue there. Sorry for clobbing this thread.

**Comment by MM-Lehmann** [comment]
Are you basing your codebase on top of `dash-core-components` for every release or per package requirements? If so, should the newest features/fixes not be available for your components as well?
For instance, https://github.com/plotly/dash/issues/1034 requested support of `Null` values to be persisted. As per my last comment on that thread, `dcc.Input` does so actually, but not `dbc.Input`. 
Could you investigate/advise, please?

**Comment by MM-Lehmann** [comment]
Thanks for the clarification. I wasn't clearing the button through a callback though, but merely hitting backspace/del on the keyboard.   
  
Also, i should have clarified, this only occurs with numeric input types (i.e. setting `type='number'`)
  
It appears `null` is not allowed in this type.
  

  

  

  
  
  
  
  
>

**Comment by tcbegley** [comment]
Null persistence for numeric inputs fixed in version 0.10.3.

**Comment by MM-Lehmann** [comment]
Cool. Thanks a bunch!

**Comment by ken19960315** [comment]
Hi, could persistence also be supported for Nav?

**Comment by tcbegley** [comment]
Hi @ken19960315 

What properties of the `Nav` do you want to persist exactly?

**Comment by ken19960315** [comment]
Hi @tcbegley,
Thanks for your reply.
Currently, I have followed this [tutorial](https://dash-bootstrap-components.opensource.faculty.ai/examples/simple-sidebar/) to create a side bar in my application, and I created a button to dynamically generate NavLink inside Nav.
However, the created NavLink would disappear if the page is refreshed.
Thus, I would like to know if Nav could have the same persistence property to keep these NavLinks.

**Comment by tcbegley** [comment]
I think persisting the children of a component is not recommended in Dash.

But you can achieve what you want to do with `dcc.Store`. Here's a little demo

```python
import re

import dash
import dash_bootstrap_components as dbc
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output, State

app = dash.Dash(external_stylesheets=[dbc.themes.BOOTSTRAP])

app.layout = html.Div(
    [
        dcc.Store(id="store", storage_type="local"),
        dbc.NavbarSimple(id="navbar", brand="Navbar"),
        dbc.Container(
            dbc.InputGroup(
                [
                    dbc.Input(id="input", type="text"),
                    dbc.InputGroupAddon(
                        dbc.Button("Add", color="primary", id="button"),
                        addon_type="append",
                    ),
                ]
            ),
            className="p-5",
        ),
    ]
)


@app.callback(
    Output("store", "data"),
    Input("button", "n_clicks"),
    [State("store", "data"), State("input", "value")],
)
def store_values(n, data, value):
    if data is None:
        data = []

    if n and value:
        slug = re.sub(r"\w+", "_", value)
        data.append({"label": value, "href": f"/{slug}"})

    return data


@app.callback(Output("navbar", "children"), Input("store", "data"))
def populate_nav(data):
    return [dbc.NavLink(item["label"], href=item["href"]) for item in data]


if __name__ == "__main__":
    app.run_server(debug=True)
```

**Comment by ken19960315** [comment]
Thanks for your solution.
It looks good to me.
